src/main/java/org/hydev/mcpm/client/database/fetcher/LocalDatabaseFetcher.java

```diff
@@ -0,0 +1,192 @@
+package org.hydev.mcpm.client.database.fetcher;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.hc.client5.http.fluent.Request;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.HttpEntity;
+import org.hydev.mcpm.client.models.Database;
+import org.hydev.mcpm.utils.HashUtils;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * A database fetcher that has a local persistent cache along and a network fallback.
+ */
+public class LocalDatabaseFetcher implements DatabaseFetcher {
+    private final URI host;
+    private final Path cacheDirectory;
+
+    private Database localDatabase;
+
+    public static final String HASH_FILE_NAME = "db.hash";
+    public static final String DATABASE_FILE_NAME = "db";
+
+    public static final String USER_AGENT = "MCPM Client";
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     * The cache directory will default to .mcpm
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     */
+    public LocalDatabaseFetcher(URI host) {
+        this(host, Path.of(".mcpm/"));
+    }
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     * @param cacheDirectory The directory that the cached database and hash file will be stored (as child files).
+     */
+    public LocalDatabaseFetcher(URI host, Path cacheDirectory) {
+        this.host = host;
+        this.cacheDirectory = cacheDirectory;
+    }
+
+    private Request requestTo(String path) {
+        return Request
+            .get(URI.create(host.toString() + "/" + path))
+            .addHeader("Host", host.getHost())
+            .addHeader("User-Agent", USER_AGENT)
+            .addHeader("Accepts", "application/json");
+    }
+
+    @Nullable
+    private String fetchHostHash() {
+        try {
+            var response = requestTo(HASH_FILE_NAME).execute();
+
+            return response.returnContent().asString().trim();
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private String fetchLocalHash() {
+        try {
+            return Files.readString(cacheDirectory.resolve(DATABASE_FILE_NAME));
```

Comment:
🤔

Did you forget to calculate the hash?

Reply:
Or should this be HASH_FILE_NAME?

Reply:
Would it be better to calculate hash on the fly instead?

Reply:
This should be `HASH_FILE_NAME` yep, this is already fixed for some reason on local :eyes:. The database file is huge and I expect reading + calculating its hash is going to be expensive. Plus, the local hash is set up to lean on the side of integrity (the user probably won't see incorrect behaviour from an out of sync hash file).

Reply:
Ok yea that makes sense